cpp/src/parquet/encoding.cc

```diff
@@ -1484,6 +1521,39 @@ class DictDecoderImpl : public DecoderImpl, virtual public DictDecoder<Type> {
   // Perform type-specific initiatialization
   void SetDict(TypedDecoder<Type>* dictionary) override;
 
+  template <typename T = Type,
+            typename = std::enable_if_t<std::is_same_v<T, ByteArrayType> ||
+                                        std::is_same_v<T, LargeByteArrayType>>>
+  void SetByteArrayDict(TypedDecoder<Type>* dictionary) {
+    DecodeDict(dictionary);
+
+    auto dict_values = reinterpret_cast<ByteArray*>(dictionary_->mutable_data());
+
+    int total_size = 0;
+    for (int i = 0; i < dictionary_length_; ++i) {
+      if (AddWithOverflow(total_size, dict_values[i].len, &total_size)) {
+        throw ParquetException("String/Binary length to large");
+      }
+    }
+    PARQUET_THROW_NOT_OK(byte_array_data_->Resize(total_size,
+                                                  /*shrink_to_fit=*/false));
+    PARQUET_THROW_NOT_OK(
+        byte_array_offsets_->Resize((dictionary_length_ + 1) * sizeof(int32_t),
```

Comment:
I am not sure I understand the question. Reading the code, we can see `total_size` will not overflow because of `AddWithOverflow`. Moreover, `ResizableBuffer::new_size` takes a `int64_t`, which a 32 bit int fits in.

Reply:
Ok I got it. This patch uses `int32_t` as offset, and maximum will not beyond `numeric_limit<int32_t>::max`. So it's also ok to use `int32_t` as offset.